torch/_inductor/lowering.py

```diff
@@ -3814,12 +3815,91 @@ def loader(idx):
     )
 
 
+def inplace_constant_pad_nd(x, padding, fill_value):
+    """
+    This optimization changes the semantics of padding from 'clone'
+    style to 'view' style.
+
+    Thanks to functionalization, this change can still maintain numerical
+    correctness.
+    """
+
+    def _padding_can_be_fused():
+        """
+        Conservatively check if padding can be fused with downstream op.
+        1. if the downstream op is a sum, then there is little benefit to
+           do inplace padding
+        2. if the downstream op is a matmul, doing inplace padding can
+           save membw.
+        """
+        current_node = V.graph.current_node
+        if current_node is None:
+            return True  # be conservative
+        users = tuple(current_node.users)
+        if len(users) == 1 and users[0].target in (
+            aten.mm.default,
+            aten.addmm.default,
+        ):
+            return False
+
+        return True  # be conservative
+
+    if _padding_can_be_fused():
+        return None
+
+    # Only handle 2D case for now
+    if len(padding) != 4 or len(x.get_size()) != 2:
+        return None
+
+    if not ir.is_storage_and_layout(x):
+        return None
+
+    x.freeze_layout()
+
+    _, layout = ir.as_storage_and_layout(x)
+    strides = layout.stride
+    if strides[1] != 1:
+        return None
```

Comment:
I'm a bit confused about this. 
if we have a tensor of size `[62, 2048]` and we are padding `[0, 2, 0, 0]` the input tensor will be allocated for size [62, 2048]. 

We can't just zero_ `[63:64, :]` because we haven't actually made that allocation. We don't own this memory. 

This is also basically kind of a rewrite of the aten.cat optimization. Which I think we should be doing for mm planning but @oniononion36 noticed that using constant_pad_nd speeds up the padding compared to cat. 		We should just fix the slowdown, and there is a repro D64571598, although i'm not sure if i'm using it correctly.

Reply:
The semantic of the padding parameter is kind of 'reversed'. The semantic of padding [0, 2, 0, 0] to a [62, 2048] tensor is to add 2 elements after each row (not adding 2 rows).

If we have extra space (due to a larger stride for dim 0), we can just inplace update that extra space.

Reply:
Potentially similar optimization can be done if we want to pad with [0, 0, 0, 2] (e.g. padding 2 rows). But I don't see workloads benefit from this so far.

Reply:
Do you know how the non-dense tensor is produced ? Why are we not just computing the constant_pad_nd shape in the forward ?

Reply:
> Do you know how the non-dense tensor is produced ? 

yea. This is generated by 'comprehensive padding' for the benefits to speed up the computation of grad_softmax_input (aka. grad_matmul_output). 

> Why are we not just computing the constant_pad_nd shape in the forward ?

Because the tensor being padded is a gradient.

In an idea world, if we are not doing mm padding in such an early stage (pattern matcher), maybe we can skip some mm padding (when we know we are doing comprehensive padding) to make the optimization looks more straightforward.